### PR TITLE
Make request timeout and TLS properties configurable.

### PR DESCRIPTION
**Pull request by suprajanarasimhan**
Related to https://github.com/wavefrontHQ/wavefront-sdk-go/issues/6.
This is an extension of PR #111, which I closed due to having introduced undesired merge commits in git history. 

**Detailed description:**
Make request timeout and TLS properties configurable. 

Add new configuration options to README.

  Signed-off-by: Devon Warshaw <warshawd@vmware.com>
  Signed-off-by: Jeanette Booher <jbooher@vmware.com>
  Signed-off-by: Supraja Narasimhan <suprajan@vmware.com>

### CODE REVIEW

**Comment by suprajanarasimhan** [comment]
Hi @oppegard and @LukeWinikates, this PR is an extension of #111 with a cleaner git history.  I explained the decision to create this PR in [this comment](https://github.com/wavefrontHQ/wavefront-sdk-go/pull/111#issuecomment-1301506714) in #111. Thank you for helping me through this as I learn.

File: README.md [hunk 1]
```diff
@@ -197,7 +197,9 @@ func main() {
         wavefront.BatchSize(20000), // Send batches of 20,000.
         wavefront.FlushIntervalSeconds(5), // Flush every 5 seconds.
         wavefront.MetricsPort(4321), // Use port 4321 for metrics.
-        wavefront.TracesPort(40001), // Use port 40001 for traces.
+        wavefront.TracesPort(40001), // Use port 40001 for traces.
```
Comment (keep94) [inline README.md:200]: We moved all the example code within the source code so that it is compiled as part of the test suite and so that this example code shows up on pkg.go.dev. Can you please resync this PR before I review?

File: internal/reporter.go [hunk 2]
```diff
@@ -18,11 +19,20 @@ type reporter struct {
 }
 
 // NewReporter create a metrics Reporter
-func NewReporter(server string, token string) Reporter {
+func NewReporter(server string, token string, timeout time.Duration, tlsConfigOptions *tls.Config) Reporter {
```
Comment (keep94) [inline internal/reporter.go:22]: Have you considered making NewReporter() look like this:
```func NewReporter(server string, token string, client *http.Client) Reporter```
Since configuration has the tls.config and the timeout you could add a ```NewClient() *http.Client``` method to configuration.

**Comment by suprajanarasimhan** [comment]
Hello @keep94, nice to meet you! Thank you for taking the time to review -- I will start addressing these today.

File: README.md (hunk 1 above)
Comment (suprajanarasimhan) [reply README.md]: Yes, done!

File: internal/reporter.go (hunk 2 above)
Comment (suprajanarasimhan) [reply internal/reporter.go]: I hadn't! Thanks, I really like this suggestion, since it reduces the number of arguments to NewReporter(). Your way of describing it made it easy to implement. 😊

**Comment by suprajanarasimhan** [comment]
> Hello @keep94, nice to meet you! Thank you for taking the time to review -- I will start addressing these today.

I made the changes and left the comments open so you can decide if they can be resolved. Thank you!

File: senders/client_factory.go [hunk 3]
```diff
@@ -230,6 +239,19 @@ func TracesPort(port int) Option {
...
+// Timeout sets the HTTP timeout (in seconds). Defaults to 10 seconds.
+func Timeout(timeout time.Duration) Option {
+	return func(cfg *configuration) {
+		cfg.Timeout = timeout
+	}
+}
+
+func TLSConfigOptions(tlsCfg *tls.Config) Option {
+	return func(cfg *configuration) {
+		cfg.TLSConfigOptions = tlsCfg
```
Comment (keep94) [inline senders/client_factory.go:251]: The caller could call TLSConfigOptions() to get an Option, and sometime later change the tlsCfg they pass in. If they do that, it will change the behavior of the Option they created too. Ideally an Option should be immutable.  Consider calling Clone() on tlsCfg to get a defensive copy Like this:  

```
func TLSConfigOptions(tlsCfg *tls.Config) Option {
    tlsCopy := tlsCfg.Clone()
    return func(cfg *configuration) {
        cfg.TLSConfigOptions = tlsCopy
    }
}
```

**Review by keep94 — COMMENTED** [review]
Thanks for the changes and the great work.  I had one more comment. Have to go, so I'll look over it again on Monday.

File: senders/client_factory.go (hunk 3 above)
Comment (suprajanarasimhan) [reply senders/client_factory.go]: Thank you, @keep94! I made the change, though I'm still trying to understand the concept you described by reading up on Golang (yesterday and today). Thought it might be worth updating you.

File: senders/client_factory.go [hunk 4]
```diff
@@ -230,6 +239,20 @@ func TracesPort(port int) Option {
...
 }
 
+// Timeout sets the HTTP timeout (in seconds). Defaults to 10 seconds.
+func Timeout(timeout time.Duration) Option {
+	return func(cfg *configuration) {
+		cfg.Timeout = timeout
+	}
+}
+
+func TLSConfigOptions(tlsCfg *tls.Config) Option {
```
[the remark this answers is not in the document]
Comment (suprajanarasimhan) [reply senders/client_factory.go]: Re: nit 
Thanks, change made!

File: internal/reporter.go [hunk 5]
```diff
@@ -17,15 +18,26 @@ type reporter struct {
...
 		token:     token,
-		client:    &http.Client{Timeout: time.Second * 10},
+		client:    client,
 	}
 }
 
+func NewClient(timeout time.Duration, tlsConfig *tls.Config) *http.Client {
+	var client *http.Client
+	if tlsConfig == nil {
+		client = &http.Client{Timeout: timeout}
```
Comment (keep94) [inline internal/reporter.go:33]: You could simplify this function like so:
```
if tlsConfig == nil {
    return &http.Client{Timeout: timeout}
}
transport := &http.Transport{TLSClientConfig: tlsConfig}
return &http.Client{Timeout: timeout, Transport: transport}
```

File: senders/client_factory.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (keep94) [reply senders/client_factory.go:249]: @LukeWinikates It is correct for the tlsCfg parameter to be a pointer. The reason being is that the tls.Config type contains a mutex, and the documentation for go mutexes explicitly prohibits assignment by value.  This is why they created the Clone() function.

File: senders/client_factory.go [hunk 6]
```diff
@@ -246,10 +246,10 @@ func Timeout(timeout time.Duration) Option {
 	}
 }
 
-func TLSConfigOptions(tlsCfg *tls.Config) Option {
+func TLSConfigOptions(tlsCfg tls.Config) Option {
```
Comment (keep94) [inline senders/client_factory.go]: Please don't pass tlsCfg by value. I looked at the tls.Config type and it contains an unexported mutex. Mutexes cannot be safely copied by assignment.

File: senders/test_server_test.go [hunk 7]
```diff
@@ -32,10 +32,10 @@ type testServer struct {
 	URL         string
 }
 
-func (s *testServer) TLSConfig() *tls.Config {
+func (s *testServer) TLSConfig() tls.Config {
 	certpool := x509.NewCertPool()
 	certpool.AddCert(s.httpServer.Certificate())
-	return &tls.Config{
```
[the remark this answers is not in the document]
Comment (suprajanarasimhan) [reply senders/test_server_test.go]: Thanks a lot, Travis! Got it -- if it contains filtered or unexported fields, pass as a pointer, not as a value.

**Comment by suprajanarasimhan** [comment]
Hello @keep94! Coming back to this after the holidays and I made the changes you requested. 

Overall, your feedback exposed me to concepts and code style that I can try to apply in the future, too! I really appreciate it. 🙏🏽

**Comment by keep94** [comment]
I will have time Friday afternoon to take a look. Is that ok?

File: internal/reporter.go (hunk 5 above)
Comment (oppegard) [reply internal/reporter.go:33]: I'm confused by the change as-is @suprajanarasimhan:
```go
	if tlsConfig == nil {
		client = &http.Client{Timeout: timeout}
	}		
	transport := &http.Transport{TLSClientConfig: tlsConfig}
	client = &http.Client{Timeout: timeout, Transport: transport}

    return client
```
The `if` condition will have no effect on `client` since it's overwritten below that. Did you mean to follow Travis' original suggestion?

```go
if tlsConfig == nil {
    return &http.Client{Timeout: timeout}
}
transport := &http.Transport{TLSClientConfig: tlsConfig}
return &http.Client{Timeout: timeout, Transport: transport}
```

Comment (oppegard) [reply internal/reporter.go:33]: @suprajanarasimhan, can we add test coverage with a nil and non-nil `tlsConfig`, to catch this logic error?

Comment (keep94) [reply internal/reporter.go:33]: As Glenn already stated, consider adding a test for internal.NewClient().  When the tlsConfig parameter is nil, the Transport field of returned http client should be nil.

**Comment by keep94** [comment]
@suprajanarasimhan I just reviewed your PR. I agree with what oppegard has said in his review, but for some reason my comments aren't showing up in this conversation page.

**Comment by suprajanarasimhan** [comment]
I will revisit this feedback later today -- thank you!

File: internal/reporter.go (hunk 5 above)
Comment (suprajanarasimhan) [reply internal/reporter.go]: Hi @oppegard, thank for catching this logic error! 🙏🏽 @ktollivervmw, @jessepye and I added the test you recommended.

**Review by oppegard — APPROVED** [review]
lgtm
